examples/train-text-from-scratch/train-text-from-scratch.cpp

```diff
@@ -679,15 +679,23 @@ struct ggml_tensor * llama_build_train_graphs(
         }
     };
 
+    // KQ_pos - contains the positions
+    struct ggml_tensor * KQ_pos = ggml_new_tensor_1d(ctx, GGML_TYPE_I32, N);
+    {
+        int * data = (int *) KQ_pos->data;
+        for (int i = 0; i < N; ++i) {
+            data[i] = n_past + i;
+        }
+    }
+
```

Comment:
@xaedes I'm changing the API of `ggml_rope` to take an entire vector with positions instead of `n_past`. I have a small concern about this particular change in `train-text-from-scratch` and cannot test it atm. I'm not sure if the allocator won't make some intermediate results to overwrite the data of `KQ_pos` at some point.

In other places, we fix this using `ggml_allocr_alloc()`:

https://github.com/ggerganov/llama.cpp/blob/1fb033fd85f8125d2830bbfe6d384be3baa17ae8/llama.cpp#L2431-L2439

But wasn't sure if it's applicable here.

Reply:
During training (finetune and train-text-from-scratch) n_past is always zero, so I guess KQ_pos would always be empty. 

> not sure if the allocator won't make some intermediate results to overwrite the data

To avoid deallocation of certain tensors T I added a temporary scale_inplace(T, 1.0f) operation at the end of the computation graph before giving it to the allocator. With this the allocator cannot deallocate T before the original end of the graph. Those temporary operations are removed from the graph after allocations are done.
For example here: https://github.com/ggerganov/llama.cpp/blob/5ce74ee4613c06bf3391c72d7115d10726200bff/examples/finetune/finetune.cpp#L768

Reply:
Hah, clever solution :) I added a scale op for `KQ_pos` to be safe.

> During training (finetune and train-text-from-scratch) n_past is always zero, so I guess KQ_pos would always be empty.

Btw, when `n_past == 0`, the `KQ_pos` tensor would have values `0, 1, 2, 3, ...` (i.e. `n_past + i`).